<a href="https://colab.research.google.com/github/LokeRuiKee/ml-dl-playground/blob/main/DCNN_experiment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Deep Convolutional Network Template Code

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                                               # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='relu'),                                             # Convolutional layer with 128 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                                           # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])


In [ ]:
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 58s 30ms/step - loss: 0.2669 - accuracy: 0.9180
Epoch 2/5
1875/1875 [==============================] - 58s 31ms/step - loss: 0.0917 - accuracy: 0.9739
Epoch 3/5
1875/1875 [==============================] - 55s 29ms/step - loss: 0.0661 - accuracy: 0.9806
Epoch 4/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0522 - accuracy: 0.9851
Epoch 5/5
313/313 [==============================] - 3s 9ms/step - loss: 0.0488 - accuracy: 0.9872
Test accuracy: 0.9872000217437744


# 7 activation, optimizer = adam

## relu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                                               # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='relu'),                                             # Convolutional layer with 128 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                                           # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 58s 30ms/step - loss: 0.2577 - accuracy: 0.9199
Epoch 2/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0837 - accuracy: 0.9761
Epoch 3/5
1875/1875 [==============================] - 59s 31ms/step - loss: 0.0612 - accuracy: 0.9820
Epoch 4/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0480 - accuracy: 0.9857
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0474 - accuracy: 0.9865
Test accuracy: 0.9865000247955322


## LeakyReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 61s 31ms/step - loss: 0.2070 - accuracy: 0.9362
Epoch 2/5
1875/1875 [==============================] - 58s 31ms/step - loss: 0.0717 - accuracy: 0.9786
Epoch 3/5
1875/1875 [==============================] - 61s 33ms/step - loss: 0.0542 - accuracy: 0.9833
Epoch 4/5
1875/1875 [==============================] - 58s 31ms/step - loss: 0.0413 - accuracy: 0.9871
Epoch 5/5
313/313 [==============================] - 4s 13ms/step - loss: 0.0414 - accuracy: 0.9880
Test accuracy: 0.9879999756813049


## elu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='elu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and elu activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='elu'),                                               # Convolutional layer with 64 filters of size 3x3 and elu activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='elu'),                                             # Convolutional layer with 128 filters of size 3x3 and elu activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='elu'),                                                           # Fully connected layer with 128 neurons and elu activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 64s 34ms/step - loss: 0.2103 - accuracy: 0.9353
Epoch 2/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0805 - accuracy: 0.9754
Epoch 3/5
1875/1875 [==============================] - 63s 33ms/step - loss: 0.0651 - accuracy: 0.9805
Epoch 4/5
1875/1875 [==============================] - 63s 34ms/step - loss: 0.0537 - accuracy: 0.9837
Epoch 5/5
313/313 [==============================] - 3s 9ms/step - loss: 0.0565 - accuracy: 0.9847
Test accuracy: 0.9847000241279602


## PReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='PReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and PReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='PReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and PReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='PReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and PReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='PReLU'),                                                           # Fully connected layer with 128 neurons and PReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 72s 38ms/step - loss: 0.2487 - accuracy: 0.9228
Epoch 2/5
1875/1875 [==============================] - 72s 38ms/step - loss: 0.0775 - accuracy: 0.9770
Epoch 3/5
1875/1875 [==============================] - 71s 38ms/step - loss: 0.0557 - accuracy: 0.9834
Epoch 4/5
1875/1875 [==============================] - 71s 38ms/step - loss: 0.0432 - accuracy: 0.9874
Epoch 5/5
313/313 [==============================] - 4s 12ms/step - loss: 0.0670 - accuracy: 0.9825
Test accuracy: 0.9825000166893005


## Sigmoid

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='sigmoid', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and PReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='sigmoid'),                                               # Convolutional layer with 64 filters of size 3x3 and sigmoid activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='sigmoid'),                                             # Convolutional layer with 128 filters of size 3x3 and sigmoid activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='sigmoid'),                                                           # Fully connected layer with 128 neurons and sigmoid activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 63s 33ms/step - loss: 2.3112 - accuracy: 0.1069
Epoch 2/5
1875/1875 [==============================] - 60s 32ms/step - loss: 1.6855 - accuracy: 0.3717
Epoch 3/5
1875/1875 [==============================] - 65s 35ms/step - loss: 0.2454 - accuracy: 0.9286
Epoch 4/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.1545 - accuracy: 0.9549
Epoch 5/5
313/313 [==============================] - 4s 11ms/step - loss: 0.1006 - accuracy: 0.9687
Test accuracy: 0.9686999917030334


## tanh

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='tanh', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and tanh activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='tanh'),                                               # Convolutional layer with 64 filters of size 3x3 and tanh activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='tanh'),                                             # Convolutional layer with 128 filters of size 3x3 and tanh activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='tanh'),                                                           # Fully connected layer with 128 neurons and tanh activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 63s 33ms/step - loss: 0.2022 - accuracy: 0.9383
Epoch 2/5
1875/1875 [==============================] - 58s 31ms/step - loss: 0.0808 - accuracy: 0.9756
Epoch 3/5
1875/1875 [==============================] - 61s 32ms/step - loss: 0.0653 - accuracy: 0.9801
Epoch 4/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.0511 - accuracy: 0.9844
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0548 - accuracy: 0.9828
Test accuracy: 0.9828000068664551


## softmax

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='softmax', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and softmax activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='softmax'),                                               # Convolutional layer with 64 filters of size 3x3 and softmax activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='softmax'),                                             # Convolutional layer with 128 filters of size 3x3 and softmax activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='softmax'),                                                           # Fully connected layer with 128 neurons and softmax activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 70s 37ms/step - loss: 2.3015 - accuracy: 0.1124
Epoch 2/5
1875/1875 [==============================] - 69s 37ms/step - loss: 2.3015 - accuracy: 0.1120
Epoch 3/5
1875/1875 [==============================] - 72s 39ms/step - loss: 2.3016 - accuracy: 0.1123
Epoch 4/5
1875/1875 [==============================] - 66s 35ms/step - loss: 2.2125 - accuracy: 0.1749
Epoch 5/5
313/313 [==============================] - 4s 13ms/step - loss: 1.2941 - accuracy: 0.5949
Test accuracy: 0.5949000120162964


## conclusion
Best activation function: LeakyReLU. Test accuracy: 0.9879999756813049

# Best activation function: LeakyReLU, 8 optimizer

## SGD

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='SGD',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 68s 36ms/step - loss: 0.9880 - accuracy: 0.6831
Epoch 2/5
1875/1875 [==============================] - 58s 31ms/step - loss: 0.2441 - accuracy: 0.9251
Epoch 3/5
1875/1875 [==============================] - 58s 31ms/step - loss: 0.1730 - accuracy: 0.9475
Epoch 4/5
1875/1875 [==============================] - 59s 31ms/step - loss: 0.1377 - accuracy: 0.9581
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0752 - accuracy: 0.9764
Test accuracy: 0.9764000177383423


## SGD with Momentum

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer=SGDm,
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 60s 31ms/step - loss: 0.2994 - accuracy: 0.9032
Epoch 2/5
1875/1875 [==============================] - 58s 31ms/step - loss: 0.0861 - accuracy: 0.9733
Epoch 3/5
1875/1875 [==============================] - 57s 30ms/step - loss: 0.0618 - accuracy: 0.9810
Epoch 4/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.0499 - accuracy: 0.9846
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0496 - accuracy: 0.9868
Test accuracy: 0.9868000149726868


## RMSProp

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='RMSProp',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 60s 31ms/step - loss: 0.2058 - accuracy: 0.9373
Epoch 2/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.0735 - accuracy: 0.9783
Epoch 3/5
1875/1875 [==============================] - 58s 31ms/step - loss: 0.0579 - accuracy: 0.9830
Epoch 4/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.0485 - accuracy: 0.9862
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0571 - accuracy: 0.9843
Test accuracy: 0.9843000173568726


## adam

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='Adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 61s 32ms/step - loss: 0.2146 - accuracy: 0.9335
Epoch 2/5
1875/1875 [==============================] - 59s 32ms/step - loss: 0.0728 - accuracy: 0.9781
Epoch 3/5
1875/1875 [==============================] - 61s 32ms/step - loss: 0.0561 - accuracy: 0.9831
Epoch 4/5
1875/1875 [==============================] - 59s 31ms/step - loss: 0.0425 - accuracy: 0.9866
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0515 - accuracy: 0.9864
Test accuracy: 0.9864000082015991


## adamax

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='Adamax',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 61s 32ms/step - loss: 0.3265 - accuracy: 0.8997
Epoch 2/5
1875/1875 [==============================] - 58s 31ms/step - loss: 0.1105 - accuracy: 0.9673
Epoch 3/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.0799 - accuracy: 0.9769
Epoch 4/5
1875/1875 [==============================] - 59s 32ms/step - loss: 0.0649 - accuracy: 0.9806
Epoch 5/5
313/313 [==============================] - 4s 13ms/step - loss: 0.0508 - accuracy: 0.9851
Test accuracy: 0.9850999712944031


## nadam

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='Nadam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 63s 32ms/step - loss: 0.2057 - accuracy: 0.9361
Epoch 2/5
1875/1875 [==============================] - 63s 34ms/step - loss: 0.0733 - accuracy: 0.9782
Epoch 3/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.0536 - accuracy: 0.9837
Epoch 4/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0443 - accuracy: 0.9859
Epoch 5/5
313/313 [==============================] - 4s 12ms/step - loss: 0.0431 - accuracy: 0.9884
Test accuracy: 0.9883999824523926


## AdaGrad

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='AdaGrad',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 59s 31ms/step - loss: 1.9409 - accuracy: 0.3734
Epoch 2/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.8666 - accuracy: 0.7395
Epoch 3/5
1875/1875 [==============================] - 59s 31ms/step - loss: 0.5329 - accuracy: 0.8439
Epoch 4/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.4126 - accuracy: 0.8795
Epoch 5/5
313/313 [==============================] - 4s 14ms/step - loss: 0.2362 - accuracy: 0.9277
Test accuracy: 0.9276999831199646


## AdaDelta

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='AdaDelta',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 65s 34ms/step - loss: 2.2959 - accuracy: 0.1285
Epoch 2/5
1875/1875 [==============================] - 61s 33ms/step - loss: 2.2726 - accuracy: 0.1836
Epoch 3/5
1875/1875 [==============================] - 60s 32ms/step - loss: 2.2469 - accuracy: 0.2405
Epoch 4/5
1875/1875 [==============================] - 62s 33ms/step - loss: 2.2161 - accuracy: 0.2958
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 2.1368 - accuracy: 0.5771
Test accuracy: 0.5770999789237976


## conclusion
Best optimizer: Nadam, Test accuracy: 0.9883999824523926

#Experimenting Best Output Activation Function


## output relu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='relu')                                                       # Output layer with 10 neurons for classification (relu activation)
])

# Compile the model
model.compile(optimizer='Nadam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 79s 41ms/step - loss: 2.4114 - accuracy: 0.0986
Epoch 2/5
1875/1875 [==============================] - 62s 33ms/step - loss: 2.3041 - accuracy: 0.0987
Epoch 3/5
1875/1875 [==============================] - 59s 32ms/step - loss: 2.3033 - accuracy: 0.0987
Epoch 4/5
1875/1875 [==============================] - 63s 33ms/step - loss: 2.3037 - accuracy: 0.0987
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 2.3026 - accuracy: 0.0980
Test accuracy: 0.09799999743700027


## output LeakyReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='LeakyReLU')                                                       # Output layer with 10 neurons for classification (LeakyReLU activation)
])

# Compile the model
model.compile(optimizer='Nadam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 62s 32ms/step - loss: 2.5679 - accuracy: 0.3167
Epoch 2/5
1875/1875 [==============================] - 63s 34ms/step - loss: 2.3055 - accuracy: 0.1448
Epoch 3/5
1875/1875 [==============================] - 65s 35ms/step - loss: 2.3030 - accuracy: 0.1442
Epoch 4/5
1875/1875 [==============================] - 65s 35ms/step - loss: 2.3029 - accuracy: 0.1328
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 2.3026 - accuracy: 0.1488
Test accuracy: 0.14880000054836273


## output elu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='elu')                                                       # Output layer with 10 neurons for classification (elu activation)
])

# Compile the model
model.compile(optimizer='Nadam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 66s 34ms/step - loss: 2.3394 - accuracy: 0.2724
Epoch 2/5
1875/1875 [==============================] - 64s 34ms/step - loss: 2.3043 - accuracy: 0.1256
Epoch 3/5
1875/1875 [==============================] - 66s 35ms/step - loss: 2.3030 - accuracy: 0.1282
Epoch 4/5
1875/1875 [==============================] - 66s 35ms/step - loss: 2.3030 - accuracy: 0.1203
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 2.3026 - accuracy: 0.1157
Test accuracy: 0.11569999903440475


## output PReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='PReLU')                                                       # Output layer with 10 neurons for classification (PReLU activation)
])

# Compile the model
model.compile(optimizer='Nadam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 64s 33ms/step - loss: 2.2562 - accuracy: 0.3835
Epoch 2/5
1875/1875 [==============================] - 60s 32ms/step - loss: 2.3067 - accuracy: 0.0996
Epoch 3/5
1875/1875 [==============================] - 60s 32ms/step - loss: 2.3030 - accuracy: 0.0991
Epoch 4/5
1875/1875 [==============================] - 62s 33ms/step - loss: 2.3034 - accuracy: 0.0989
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 2.3026 - accuracy: 0.0980
Test accuracy: 0.09799999743700027


## output sigmoid

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='sigmoid')                                                       # Output layer with 10 neurons for classification (sigmoid activation)
])

# Compile the model
model.compile(optimizer='Nadam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 61s 31ms/step - loss: 0.2073 - accuracy: 0.9360
Epoch 2/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.0729 - accuracy: 0.9783
Epoch 3/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0566 - accuracy: 0.9830
Epoch 4/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0439 - accuracy: 0.9868
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0463 - accuracy: 0.9870
Test accuracy: 0.9869999885559082


## output tanh

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='tanh')                                                       # Output layer with 10 neurons for classification (tanh activation)
])

# Compile the model
model.compile(optimizer='Nadam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 65s 33ms/step - loss: 2.2530 - accuracy: 0.2500
Epoch 2/5
1875/1875 [==============================] - 62s 33ms/step - loss: 2.0639 - accuracy: 0.2898
Epoch 3/5
1875/1875 [==============================] - 61s 33ms/step - loss: 2.4021 - accuracy: 0.2904
Epoch 4/5
1875/1875 [==============================] - 61s 33ms/step - loss: 2.3094 - accuracy: 0.1020
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 2.3026 - accuracy: 0.0980
Test accuracy: 0.09799999743700027


## output softmax

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the deep CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                                               # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Conv2D(128, (3, 3), activation='LeakyReLU'),                                             # Convolutional layer with 128 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                                                         # Max pooling layer with pool size 2x2
    Flatten(),                                                                                                # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                                           # Fully connected layer with 128 neurons and LeakyReLU activation
    Dropout(0.5),                                                                                        # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                                       # Output layer with 10 neurons for classification (softmax activation)
])

# Compile the model
model.compile(optimizer='Nadam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 62s 32ms/step - loss: 0.2051 - accuracy: 0.9369
Epoch 2/5
1875/1875 [==============================] - 61s 33ms/step - loss: 0.0737 - accuracy: 0.9771
Epoch 3/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.0573 - accuracy: 0.9832
Epoch 4/5
1875/1875 [==============================] - 66s 35ms/step - loss: 0.0459 - accuracy: 0.9863
Epoch 5/5
313/313 [==============================] - 4s 11ms/step - loss: 0.0483 - accuracy: 0.9852
Test accuracy: 0.9851999878883362


## conclusion
Best output activation function: sigmoid, Test accuracy: 0.9869999885559082